### PR TITLE
[query] Enable use of aggregators on array expressions

### PR DESCRIPTION
**Pull request by tpoterba**
This method is useful for accessing functionality that exists in the aggregator library but not the basic expression library, for instance, `call_stats`.

### CODE REVIEW

File: hail/python/hail/expr/aggregators/aggregators.py [hunk 1]
```diff
@@ -238,6 +238,36 @@ def context(self):
             return 'agg'
 
 
+def aggregate_local_array(array, f):
```
Comment (patrick-schultz) [inline hail/python/hail/expr/aggregators/aggregators.py:241]: We had discussed adding a `CollectionExpression.aggregate` method, similar to `fold`. Did you change your mind?

**Review by patrick-schultz — CHANGES_REQUESTED** [review]

File: hail/python/hail/expr/aggregators/aggregators.py (hunk 1 above)
Comment (tpoterba) [reply hail/python/hail/expr/aggregators/aggregators.py:241]: nope, forgot over the weekend amid the presentation rush. Thanks for the reminder.

File: hail/python/hail/ir/ir.py [hunk 2]
```diff
@@ -2074,6 +2074,78 @@ def renderable_bindings(self, i, default_value=None):
...
 
+class StreamAgg(IR):
+    @typecheck_method(a=IR, value_name=str, body=IR)
+    def __init__(self, a, value_name, body):
+        a = a.handle_randomness(body.uses_randomness)
+        if body.uses_randomness:
+            tuple, uid, elt = unpack_uid(a.typ)
+            body = AggLet(value_name, elt, body)
+            body = with_split_rng_state(body, uid)
+            value_name = tuple
```
[the remark this answers is not in the document]
Comment (patrick-schultz) [reply hail/python/hail/ir/ir.py]: Thanks Chris. That's almost right, but we still need to bind the `elt` (`body` expects `value_name` to be an element without uid).  And lets keep using `with_split_rng_state` to keep that logic in one place:
```
             tup, uid, elt = unpack_uid(a.typ)
             body = AggLet(value_name, elt, body)
             if body.uses_value_randomness:
                 body = Let('__rng_state', RNGStateLiteral(), body)
             if body.uses_agg_randomness(is_scan=False):
                 body = with_split_rng_state(body, is_scan=False)
             value_name = tup
```

[the remark this answers is not in the document]
Comment (chrisvittal) [reply hail/python/hail/ir/ir.py]: thank you, should be fixed now.

**Review by patrick-schultz — APPROVED** [review]

**Review by patrick-schultz — CHANGES_REQUESTED** [review]
sorry, realized this still needs a test for the randomness handling, I can add that.

File: hail/python/hail/ir/ir.py [hunk 3]
```diff
@@ -2077,14 +2077,11 @@ def renderable_bindings(self, i, default_value=None):
...
+        a = a.handle_randomness(body.uses_agg_randomness)
+        if body.uses_agg_randomness:
             tup, uid, elt = unpack_uid(a.typ)
-            body = AggLet(value_name, elt, body)
-            if body.uses_value_randomness:
-                body = Let('__rng_state', RNGStateLiteral(), body)
-            if body.uses_agg_randomness(is_scan=False):
-                body = with_split_rng_state(body, is_scan=False)
+            body = AggLet(value_name, elt, body, is_scan=False)
+            body = with_split_rng_state(body, uid, is_scan=False)
```
Comment (patrick-schultz) [inline hail/python/hail/ir/ir.py]: The test caught some problems with this. Most substantially, we shouldn't be making a new root `RNGStateLiteral` for any value randomness. That means if this occurs inside some other looping context, value randomness will be identical on each outer iteration. Instead, we should just use the rng state bound in the outer context, which means we don't need to do anything here.

Comment (chrisvittal) [reply hail/python/hail/ir/ir.py]: Thanks for explaining. Clears up why this was necessary.

**Review by chrisvittal — APPROVED** [review]

**Pull request merged**